subsys/bluetooth/host/Kconfig

```diff
@@ -344,7 +344,7 @@ config BT_USER_DATA_LEN_UPDATE
 	  info is available in the connection info.
 
 config BT_AUTO_DATA_LEN_UPDATE
-	bool "Auto-initiate Data Length Update procedure"
+	bool "Auto-initiate Data Length Update procedure" if !BT_STM32WB0
 	depends on BT_DATA_LEN_UPDATE
```

Comment:
Changes to this file should not be required, based on the changes to `soc/st/stm32/stm32wb0x/Kconfig.defconfig`.

Reply:
As this config has prompt, it is user configurable. So, if in an application, like BT shell, it is activated in `prj.conf` then the default value will be ignored as well as changes in `soc/st/stm32/stm32wb0x/Kconfig.defconfig`.

Reply:
Ok, I see. Then is is the case as well for bunch of these options (`BT_HCI_ACL_FLOW_CONTROL` for instance) and yet this is never used. Anyway, I let this to subsystem maintainers.

Reply:
What's the reason these options must remain disabled for this controller? Is it because the controller doesn't support them? Does it incorrectly report support for them when querying its features over HCI? (the host shouldn't initiate such procedures if the controller reports that it doesn't support them).

Either way, having HCI driver specific references in the host Kconfig is not the right way to deal with this. What the right/best way is depends on the answers to my earlier questions.

Reply:
Zephyr host starts several BLE procedures in parallel during connection (`perform_auto_initiated_procedures` in `subsys/bluetooth/host/conn.c`); however, ST controller does not support this "parallelism" and according to the ST Bluetooth stack experts, host should not initiate a new procedure before previous one is completed. In such case, ST controller returns a "controller busy" error (0x3A) which is not managed by the host and it's only reported on the console.
We may request further clarification in the future, but for the time being to avoid showing console error, I disabled CONFIG_BT_AUTO_PHY_UPDATE and CONFIG_BT_AUTO_DATA_LEN_UPDATE.

By the way, I will remove any changes in the host part added for handling special use cases to minimize the impact on any other piece of code.

Reply:
If you can show where in the core spec it clearly states that these shouldn't be done in parallel, then I think we should fix the host to do them serially. If there's no such place in the spec (or it's ambiguous) then the next option is to introduce a new quirk that the HCI driver sets in its DT binding, and that then causes the host to do these things serially rather than in parallel.

What you're now doing is a workaround which I don't think we should accept, i.e. you're disabling the automated procedure when the problem is not the procedure itself rather the parallelism of the procedure together with other procedure.

@alwa-nordic any thoughts on this?

Reply:
The point in the spec is: Volume 6, part B, Section 5.

> For all other Link Layer control
> procedures, only one Link Layer control procedure shall be initiated in the Link
> Layer at a time per connection per device. A new Link Layer control procedure
> shall not be initiated until any previous Link Layer control procedure initiated by
> the same device on the same connection has completed.
> 

ST has implemented the "shall not be initiated" by returning the code 0x3A (CONTROLLER_BUSY) to host.
Now, this is a requirement in the link layer part, so I am not sure whether the host should take into account this constraint in the link layer or assume that link layer should be able to "buffer" parallel requests and execute them in sequence.

Reply:
@cvinayak @alwa-nordic seems like we likely need to fix the host side for this (see above)

Reply:
Discussion from the Bluetooth WG:

 - Quirk could be the way to go
 - The spec restrictions are on parallel control procedures, controller can still queue them up internally (i.e. not return "busy")
 - Detecting completion might not be reliable, e.g. if there was no change to data length then there might not be any event to indicate this. The quirk could require the controller to always send events to indicate control procedure completion.

@HoZHel would you be able to implement a new quirk? We could call it e.g. "serial-control-proc".

Reply:
> The point in the spec is: Volume 6, part B, Section 5.
> 
> > For all other Link Layer control
> > procedures, only one Link Layer control procedure shall be initiated in the Link
> > Layer at a time per connection per device. A new Link Layer control procedure
> > shall not be initiated until any previous Link Layer control procedure initiated by
> > the same device on the same connection has completed.
> 
> ST has implemented the "shall not be initiated" by returning the code 0x3A (CONTROLLER_BUSY) to host. Now, this is a requirement in the link layer part, so I am not sure whether the host should take into account this constraint in the link layer or assume that link layer should be able to "buffer" parallel requests and execute them in sequence.

I disagree with this interpretation. The text says it's the Link Layer that shall not start concurrent control procedures. Control procedures are not a concept the Host "knows about", and this requirement is written in the Link Layer specification.

On the practical side: How can the Host know a control procedure is completed? There is no event for this. The controller can even start control procedures by its own volition. I just don't know how we could model this in the Host.

Reply:
@HoZHel @alwa-nordic makes a fair point, and we haven't heard of any other controllers having an issue with this. Is it an option that you could get your controller firmware updated? If it's not, the quirk option is still available, since one of the reasons we have quirks is for "broken controllers".

Reply:
If we do the quirk: I suggest we do it defensive programming-style, with a delayed retry. The delay should be timeout-based, but optionally some events can reduce or skip the timeout.

Reply:
@alwa-nordic @jhedberg, I was reporting the opinion of our BLE stack team people and I will try to involve them in the conversation. On the other hand, I suspect it could be a fairly long discussion. Our target was to have this PR approved before next code freeze. Is there anyway we could push this PR for approval ? or you prefer to clean the controller busy with a quirk or a 
 discussion with host guys ?

Reply:
> If we do the quirk: I suggest we do it defensive programming-style, with a delayed retry. The delay should be timeout-based, but optionally some events can reduce or skip the timeout.

@alwa-nordic, you mean to implement the quirk at driver level which should implement a delayed retry when controller busy is detected? Or you mean that host part should be changed to make a delayed retry when controller busy is returned? I would reckon that last approach would extend the host to properly handle controller busy cases.

Reply:
I am fine with this PR adding an override of the `default y`. You should also consider adding `BUILD_ASSERT` in your driver to deny these configs completely.

Aside: I even think we should consider globally removing the `default y` globally. I think explicit is better than implicit, and especially as new features are introduced, it's safer not to automatically add extra effects. We could have snippets or a config for "recommended getting-started config", that selects "most" features and automatics, but is still explicitly asked for by the application developer.

Reply:
> > If we do the quirk: I suggest we do it defensive programming-style, with a delayed retry. The delay should be timeout-based, but optionally some events can reduce or skip the timeout.
> 
> @alwa-nordic, you mean to implement the quirk at driver level which should implement a delayed retry when controller busy is detected? Or you mean that host part should be changed to make a delayed retry when controller busy is returned? I would reckon that last approach would extend the host to properly handle controller busy cases.

Either in the Host or the driver is fine by me. The driver could implement this with vendor-specific events, so it might be more performant. OTOH, in the Host, defensive programming is never a bad idea.

Reply:
> @alwa-nordic @jhedberg, I was reporting the opinion of our BLE stack team people and I will try to involve them in the conversation. On the other hand, I suspect it could be a fairly long discussion. Our target was to have this PR approved before next code freeze. Is there anyway we could push this PR for approval ? or you prefer to clean the controller busy with a quirk or a discussion with host guys ?

Optimally, the fix should be in the Controller. So if we could convince them, that would be great. I welcome the discussion.